### PR TITLE
Add Janitor Dockerfile

### LINKED ISSUE #286: Janitor Support
Hello everyone!

You can contribute to dspace-angular in a breeze using [Janitor](https://janitor.technology) today! It makes your work easier by giving you access to a ready made environment to work on the project. 

Currently, the dspace-angular environment is made using Docker image (think of it like a sandbox) and is found under the Janitor Technology [collection](https://hub.docker.com/r/janitortechnology/dspace/) of images. I think it would be great to decentralize this by bringing the Dockerfile (a used to make the Docker image) to this repository. The following outlines the requirements for Janitor: https://discourse.janitor.technology/t/how-to-add-a-project/165/3

I would love to hear your thoughts on this addition?

For more information on DSpace's Docker initiative, see: https://dspace-labs.github.io/DSpace-Docker-Images/

cc @jankeromnes, @terrywbrady

### PR DESCRIPTION
**Pull request by PTrottier**
Fixes #286 

This PR provides the structure for adding DSpace-Angular to Janitor with a Dockerfile similar to the one used currently by Janitor. To ensure compatibility with Janitor, we elected to rebuild a new Dockerfile instead of basing the code on the existing.

cc @jankeromnes, @terrywbrady, @artlowel

### CODE REVIEW

File: .janitor/dspace-update.dockerfile [hunk 1]
```diff
@@ -0,0 +1,13 @@
+FROM janitortechnology/dspace
```
Comment (terrywbrady) [inline .janitor/dspace-update.dockerfile]: What is the purpose of this Dockerfile?  How does it relate to the other file?

Since this file is performing a git fetch within the image, does this file need to reside in DSpace-angular?

File: .janitor/dspace.dockerfile [hunk 2]
```diff
@@ -0,0 +1,25 @@
+FROM janitortechnology/ubuntu-dev
+
+# Add the project 
+ADD ../ .
```
Comment (terrywbrady) [inline .janitor/dspace.dockerfile]: Have you been able to confirm that the .dockerignore is honored in this operation?

File: .janitor/dspace.dockerfile [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+WORKDIR /home/user/dspace-angular/
+
+# Add server configuration
+COPY --chown=user:user ./home/user/dspace
+COPY environment.prod.js /home/user/dspace-angular/config/
```
Comment (terrywbrady) [inline .janitor/dspace.dockerfile]: This operation is binding all of the configuration settings into the image build.  Would it be possible to allow the user to specify these at launch time instead?

Take a look at the approach that I used for the DSpace 7 compose file.  The environmental overrides are supplied when the compose file is started rather than when the image is built.

https://github.com/DSpace-Labs/DSpace-Docker-Images/tree/master/docker-compose-files/dspace7-compose

Would this be possible within Janitor?

File: .janitor/janitor.json [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        "openPort": "3000"
+      },
+      "Install global dependencies": "yarn run global",
+      "Install local dependencies": "yarn install",
+      "Start": {
+        "cmd": "yarn start",
+        "openPort": "3000"
+      },
+      "Clean": "yarn run clean",
+      "Update source code": "git pull --rebase origin master",
```
Comment (terrywbrady) [inline .janitor/janitor.json]: The image has been built from a branch of source code, yet this operation will update source code within the image.  An explanation of how this relates to the built image would be useful.

File: .janitor/dspace.dockerfile (hunk 2 above)
Comment (PTrottier) [reply .janitor/dspace.dockerfile]: Not yet, will do some testing and report back

File: .janitor/janitor.json (hunk 4 above)
Comment (PTrottier) [reply .janitor/janitor.json]: That's a command a user can run to update the source code, useful for rebasing to changes from master.

File: .janitor/dspace-update.dockerfile (hunk 1 above)
Comment (PTrottier) [reply .janitor/dspace-update.dockerfile]: I am not sure, but I think it would be a good idea to keep everything together.

File: .janitor/dspace.dockerfile (hunk 3 above)
Comment (PTrottier) [reply .janitor/dspace.dockerfile]: In Janitor, the environments are pre-built so that they can be started in 2.1 seconds, so I am not sure this is possible. I will investigate and report back. 

cc @jankeromnes

Comment (jankeromnes) [reply .janitor/dspace.dockerfile]: Janitor doesn't support Docker compose (yet). However, environment configuration can be done in three different ways:

- Project-specific configuration files (or ENV variables), included in the project image: useful for default configurations that will be useful for most users, yet these can still be tweaked as much as needed by users once their containers are running
- User-specific configuration files, managed across all the user's containers: these files are generated from [templates](https://github.com/JanitorTechnology/janitor/tree/master/templates/configurations), editable in [Settings](https://janitor.technology/settings/), and deployed in *all* the user's containers (useful to configure common tools, regardless of which project they're working on)
- Ad-hoc commands or scripts run by users in their containers: these scripts can be added to `janitor.json`'s `"scripts"` section for convenience (e.g. the Discourse image has a `bundle exec rake admin:create` script that users can decide to run, either manually or by using the ["Create admin user" IDE script](https://github.com/JanitorTechnology/dockerfiles/blob/9ee71d49efce9f5a7d875c9a82d9f7e9e54d9c19/discourse/janitor.json#L39) in one click)

File: .janitor/dspace-update.dockerfile (hunk 1 above)
Comment (jankeromnes) [reply .janitor/dspace-update.dockerfile]: Such an `*-update.dockerfile` can be used to build incremental image updates (as opposed to full image rebuilds, which may take more disk space). This technique was used to upgrade project images while saving disk space in our servers, but it is not very useful for relatively small projects (image < 30GB).

TL;DR: Feel free to remove this file entirely.

**Comment by tdonohue** [comment]
Closing as obsolete
